radixdlt-core/radixdlt/src/test/java/com/radixdlt/network/p2p/PendingOutboundChannelsManagerTest.java

```diff
@@ -0,0 +1,131 @@
+ * shall be liable for any direct or indirect, special, incidental, consequential or other
+ * losses of any kind, in tort, contract or otherwise (including but not limited to loss
+ * of revenue, income or profits, or loss of use or data, or loss of reputation, or loss
+ * of any economic or other opportunity of whatsoever nature or howsoever arising), arising
+ * out of or in connection with (without limitation of any use, misuse, of any ledger system
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.network.p2p;
+
+import static com.radixdlt.utils.TypedMocks.rmock;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import com.radixdlt.crypto.ECKeyPair;
+import com.radixdlt.environment.EventDispatcher;
+import com.radixdlt.environment.ScheduledEventDispatcher;
+import com.radixdlt.network.p2p.transport.PeerChannel;
+import com.radixdlt.network.p2p.transport.PeerOutboundBootstrap;
+import com.radixdlt.properties.RuntimeProperties;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+import org.json.JSONObject;
+import org.junit.Test;
+
+public final class PendingOutboundChannelsManagerTest {
+  @Test
+  public void
+      pending_outbound_channels_manager_should_return_the_same_future_when_called_from_diff_threads()
```

Comment:
Can you add a comment or java doc explaining how this test arose, and why it's important that the same future is returned from different threads?

I thought the issue was a combination of:
* Resolving a future during a lock
* That the future's callback was being executed on the wrong thread? (the P2PManagerThread instead of the Outbound message processing thread?)

I'm not sure it's obvious to me how this test shows that either issue is fixed? A comment might help :)

Reply:
The test as written doesn't obviously prove to me the code as a whole can't deadlock - maybe this implication is indirect?

I think it'd be useful to capture an explanation of the implication / indirection into the codebase, so someone reviewing it in the future can understand why this test is important when they consider whether to refactor/fix/delete it.

Reply:
ah, you've already answered @diegomrsantos , missed that :)

Reply:
As we aren't directly using locks, probably, we don't need to care about deadlocks anymore. The critical issue is the race condition that I believe the previous lock was trying to solve. The test tries to show it doesn't happen when replacing the lock by a `ConcurrentHashMap`. It could be made more explicit with comments.

Reply:
added a comment